src/box/lua/iproto.c

```diff
@@ -268,16 +305,115 @@ push_iproto_protocol_features(struct lua_State *L)
 					  lengthof(features));
 }
 
+/**
+ * Encodes a packet header/body argument to MsgPack: if the argument is a
+ * string, then no encoding is needed — otherwise the argument must be a Lua
+ * table. The Lua table is encoded to MsgPack using IPROTO key translation
+ * table.
+ * In both cases, the result is stored on the fiber region.
+ */
+static const char *
+encode_packet(struct lua_State *L, int idx, size_t *mp_len)
+{
+	int packet_part_type = lua_type(L, idx);
+	struct region *gc = &fiber()->gc;
+	if (packet_part_type == LUA_TSTRING) {
+		const char *arg = lua_tolstring(L, idx, mp_len);
+		char *mp = xregion_alloc(gc, *mp_len);
+		return memcpy(mp, arg, *mp_len);
+	}
+	assert(packet_part_type == LUA_TTABLE);
+	/*
+	 * Can throw a Lua exception, which we cannot catch, and cause the fiber
+	 * region to leak.
+	 */
+	struct mpstream stream;
+	mpstream_init(&stream, gc, region_reserve_cb, region_alloc_cb,
+		      luamp_error, L);
+	size_t used = region_used(gc);
+	/*
+	 * Can throw a Lua exception, which we cannot catch, and cause the fiber
+	 * region to leak. This issue is tracked in #7939.
```

Comment:
Why can't you use `lua_pcall()` / `lua_cpcall()`? C-function push is slow, but it is easily workarounded - see `luaT_tuple_encode_table_ref` as an example.

Reply:
Not only is the C function push slow, setting up the function call may also affect performance, this needs to be evaluated.

Reply:
This is why I said you have to look at `luaT_tuple_encode_table_ref`, which doesn't have the problem of slow push of C functions.

Reply:
As I said, **push of C function is not the only problem** I see here: I am also concerned with how setting up the function call (i.e., pushing arguments to Lua stack, setting up the protected environment for the call) can affect performance.

Reply:
Ah, about pushing the args - well, you don't have much choice here. Lua can throw (any push function can throw, in fact). So you have to call it safely. Either as a pushed C func or as a Lua ref to a C func.

As for the discussions/benches - they were done in notion/telegram/vk-teams/emails, I don't have the links (that is why it is important to keep stuff in commit messages). But you should be able to use `git blame` on the usecases:

1. `src/box/lua/tuple.c`
```C
/**
 * <luaT_tuple_encode_table>() reference in the Lua registry.
 *
 * Storing of the reference allows to don't create a new GCfunc
 * object each time we call the function in the protected mode.
 * It reduces Lua GC pressure in comparison with calling of
 * <lua_cpcall>() or <lua_pushcfunction>() on each invocation.
 */
static int luaT_tuple_encode_table_ref = LUA_NOREF;
```

2. An experimental patch from @igormunkin (https://github.com/tarantool/tarantool/commit/8fe7f0075c07dca4f091f606193e8d3700f3c45d) which he wanted to bench. Most likely the benchmark results are left somewhere in our patch emails.

3. https://github.com/tarantool/tarantool/commit/e88c0d21ab765d4c53bed2437c49d77b3ffe4216 - optimization of Lua stored function calls.

4. `src/box/lua/net_box.c`
```C
/**
 * We keep a reference to each C function that is frequently called with
 * lua_call so as not to create a new Lua object each time we call it.
 */
static int luaT_netbox_request_iterator_next_ref = LUA_NOREF;
```

5. `src/box/lua/watcher.c`
```C
/**
 * We keep a reference to each C function that is often called with lua_pcall
 * so as not to create a new Lua object each time we call it.
 */
static int lbox_watcher_run_lua_ref = LUA_NOREF;
```

6. https://github.com/tarantool/tarantool/commit/604b98ee01f41c4c7ff78b74fb594a438664f6e8 - speed up of `luaT_newthread()`.

As for Vladimit being unaware of this - I highly doubt it 😉. A couple of places above are done by him.

Reply:
I suggest we move this discussion to #7939 to not block this patch. But if you insist, I can bring the fix right away.

Reply:
Personally I do not care too much, I have already approved the PR. I leave it to other reviewers to decide whether we want to submit a known memory leak.

Reply:
Too many mentions, so I can't just skip this discussion. Though, I'm quite late for this PR, I just leave some links for the history: this is the [thread](https://lists.tarantool.org/tarantool-patches/cover.1592584372.git.imun@tarantool.org/) @Gerold103 is talking about, and this is the [benchmark](https://gist.github.com/igormunkin/c941074fa9fdf0f7a4f068498fb5e24c) (however, I'm not sure whether it is relevant for your case).